### ISSUE TITLE
Testing Framework

### ISSUE THREAD

**Issue by 0xcaff** [trufflesuite/truffle#500]
Will truffle ever allow testing contracts in another testing framework or use a different assertion library with `truffle test`? I really like `jest`'s snapshot testing and it's BDD style assertions. When writing contract tests, I often find myself setuping up contract state, running some oprations and comparing the expected state to the actual state. This seems like the perfect job for snapshot testing.

I know this can be accomplished using `truffle-contract` and writing tests without using truffle directly but then the `artifact` global along with the standard `truffle test` convention is lost.

**Comment by adrianmcli** [comment]
I also think this is worth considering. Perhaps we could make it possible to plug-in your own testing library in the future?

**Comment by code-chimp** [comment]
Would it be possible to at least enable Chai's [BDD style assertions](http://chaijs.com/api/bdd/)?

**Comment by EdsonAlcala** [comment]
I have seem Jest and I like the Immersive Watch Mode approach, is there a way to do that in truffle or mocha?

**Comment by Mischi** [comment]
We just released a first version of https://github.com/AsureFoundation/truffle-jest. It allows you to run your truffle tests with *Jest* instead of `truffle test`. Give it a try if you like. This is mostly a PoC and has some rough edges 😉

We played with *Mocha / Mocha CLI* too but in the end *Jest* had better support for "Async Initialization before executing tests" as *Mocha*. 

PS: I just stumbled upon `truffle-solidity-loader` and wondering if it can be used together with *Jest* to rerun the tests once  the smart contracts changed ...

Cheers,
Fabian
